src/createSlice.ts

```diff
@@ -106,8 +111,21 @@ type PrepareActionForReducer<R> = R extends { prepare: infer Prepare }
   ? Prepare
   : never
 
-type CaseReducerActions<CaseReducers extends SliceCaseReducers<any, any>> = {
-  [Type in keyof CaseReducers]: IfIsEnhancedReducer<
+type ActionForReducer<R, S> = R extends (
```

Comment:
As you don't really care for the State type here, I believe you can just leave it out. 
```typescript
type ActionForReducer<R> =
  R extends (  state: any,  action: PayloadAction<infer P> ) => any
  ? PayloadAction<P>
  : R extends {      reducer(state: any, action: PayloadAction<infer P>): any    }
  ? PayloadAction<P>
  : unknown
```

Reply:
Yeah, that's basically what I originally had.  @nickmccurdy suggested I might as well use `State` instead of `any` for consistency, and it doesn't seem to actually hurt anything.

Reply:
It doesn't hurt - it's just one more thing to carry around :)

Explanation why I think it's not necessary here:
What you are doing here is shape-matching. You're either looking for an object with a `reducer` method or for a method itself, both methods with at least two arguments. From that moment on you just care about the type of the second argument.
Sure, you can choose to only match if the first argument is `State`, but that means a second parameter to the generic, also caring about that second parameter when calling. So, essentially: more code, without any direct benefit. 
So I'd just let it out. But if you choose to let it in there, don't forget to do the same check in line 120, too.